src/rgw/rgw_data_sync.h

```diff
@@ -390,7 +390,8 @@ class RGWRemoteDataLog : public RGWCoroutinesManager {
   int init_sync_status(int num_shards);
   int run_sync(int num_shards);
 
-  void wakeup(int shard_id, set<string>& keys);
+  template<typename T>
+  void wakeup(int shard_id, T& keys);
```

Comment:
it looks like you're trying to templatize this stuff so that we can call it with either `map<int, set<rgw_data_notify_entry>>` from the v2 api or `map<int, set<string>>` from v1. but i think we should stick with the `rgw_data_notify_entry` version everywhere, and only deal with the old format inside the json logic

i pushed a commit to https://github.com/cbodley/ceph/commits/wip-pr-39477-json that adds the backward-compatible json encode and decode with `struct rgw_data_notify_v1_encoder` and `rgw_data_notify_v1_decoder`

with `rgw_data_notify_v1_decoder`, your v1 `RGWOp_DATALog_Notify` could look like this:
```c++
  bc::flat_map<int, bc::flat_set<rgw_data_notify_entry>> updated_shards;
  try {
    // decode in v1 format
    auto decoder = rgw_data_notify_v1_decoder{updated_shards};
    decode_json_obj(decoder, &p);
  }
```

with `rgw_data_notify_v1_encoder`, your fallback in `RGWDataNotifierManager::notify_all()` would use `rgw_data_notify_v1_encoder` as the template parameter for `RGWPostRESTResourceCR`.
```c++
using PostNotify1 = RGWPostRESTResourceCR<rgw_data_notify_v1_encoder, int>;
auto encoder = rgw_data_notify_v1_encoder{shards};
call(new PostNotify1(cct, conn, http, "/admin/log", pairs, encoder, nullptr));
```
and your v2 request would look like:
```c++
using PostNotify2 = RGWPostRESTResourceCR<bc::flat_map<int, bc::flat_set<rgw_data_notify_entry>>, int>;
call(new PostNotify2(cct, conn, http, "/admin/log", pairs, shards, nullptr));
```